functions/src/log-swipe.ts

```diff
@@ -0,0 +1,38 @@
+import { firestore as adminFirestore } from 'firebase-admin'
+import { z } from 'zod'
+import { APIError, AuthedUser, newEndpoint, validate } from './api'
+import { removeUndefinedProps } from '../../common/util/object'
+
+const firestore = adminFirestore()
+
+const bodySchema = z.object({
+  contractId: z.string(),
+  amount: z.number().optional(),
+  outcome: z.enum(['YES', 'NO', 'SKIP']),
+})
+
+export const logswipe = newEndpoint({}, async (req: any, auth: AuthedUser) => {
+  const { contractId, amount, outcome } = validate(bodySchema, req.body)
+  const swipeTime = Date.now()
+
+  return await firestore.runTransaction(async (trans) => {
+    const contractDoc = firestore.doc(`contracts/${contractId}`)
+    const userDoc = firestore.doc(`private-users/${auth.uid}`)
+    const [contractSnap, userSnap] = await trans.getAll(contractDoc, userDoc)
+    if (!contractSnap.exists) throw new APIError(400, 'Contract not found.')
+    if (!userSnap.exists) throw new APIError(400, 'User not found.')
+
+    const ref = firestore
+      .collection(`/private-users/${auth.uid}/seenMarkets`)
```

Comment:
I'm not sure if this should be called seen markets unless there are plans to log visits to markets here as well?

Actually, I like this collection, keyed by contract id for general use.

It could be reasonable to set `visited: true` and `lastVisitTime` on the doc `seenMarkets/{contractId}` if you navigate to a market page (or create the doc if it doesn't exist).

Either plan for this broader use case, or change the collection name to `swipes`.

Reply:
leaning towards broader use case. though you don't need `visited: true`, just the timestamp will do